pandas/lib.pyx

```diff
@@ -997,7 +1001,9 @@ cpdef ndarray[object] astype_str(ndarray arr):
         ndarray[object] result = np.empty(n, dtype=object)
 
     for i in range(n):
```

Comment:
We don't need an explicit check because we created this array with np.empty so it is mutable. The `unsafe` version of the function is the one that doesn't do the writeable check so this does what you describe unless I am missing something.